onnxruntime/test/providers/cpu/tensor/cast_op_test.cc

```diff
@@ -189,5 +208,65 @@ TEST(CastOpTest, ToString) {
   TestCastOp(gsl::make_span(int_16_input), gsl::make_span(int_string_data), shape);
 }
 
+template <typename F8>
+void CastOpTestFloat8(Saturate saturate) {
+  ASSERT_NE(saturate, Saturate::None);
+  const std::vector<int64_t> shape{2, 2, 2};
+  const std::vector<float> float_input = {NAN, -1.f, 0.0391877927f, 0.296140194f, -0.120196559f, 5.0f,
+                                          -std::numeric_limits<float>::infinity(),
+                                          std::numeric_limits<float>::infinity()};
+
+  // float output precision is 8, so the expected output differs slightly from the input due to that
+  std::vector<Float8E4M3FN> output;
+  output.reserve(float_input.size());
+  for (size_t i = 0; i < float_input.size(); ++i) {
+    output.emplace_back(Float8E4M3FN(float_input[i], saturate == Saturate::True));
+  }
+  TestCastOp<float, Float8E4M3FN>(gsl::make_span(float_input), gsl::make_span(output), shape, OpTester::ExpectResult::kExpectSuccess, "", 19, saturate);
+
+  const std::vector<MLFloat16> float16_input =
+      CastedValues<float, MLFloat16>(gsl::make_span(float_input));
+
+  TestCastOp<MLFloat16, Float8E4M3FN>(gsl::make_span(float16_input), gsl::make_span(output), shape, OpTester::ExpectResult::kExpectSuccess, "", 19, saturate);
+}
+
+TEST(CastOpTest, ToFloat8E4M3FN) {
+  int min_cuda_architecture = 11080;
+  bool enable_cuda = (nullptr != DefaultCpuExecutionProvider().get()) && HasCudaEnvironment(min_cuda_architecture);
+  bool enable_cpu = (nullptr != DefaultCpuExecutionProvider().get());
```

Comment:
is CPU EP ever not enabled?

Reply:
I replaced DefaultCpuExecutionProvider by DefaultCudaExecutionProvider. I can see many lines like this in the test `if (nullptr != DefaultCpuExecutionProvider().get()) {`.

Reply:
> I can see many lines like this in the test `if (nullptr != DefaultCpuExecutionProvider().get()) {`.

not sure if those checks are necessary.

https://github.com/microsoft/onnxruntime/blob/a8d9b29cd28999a2d0ec69c1a75e98f4b8828668/onnxruntime/test/util/default_providers.cc#L16-L23

in the current implementation, it dereferences the pointer without checking if it is null.

we can clean it up in a separate change though. maybe `DefaultCpuExecutionProvider()` should return `not_null<unique_ptr<IExecutionProvider>>`.